xSPFarmSolution resource.

wulfland opened the pull request:
I use DSC to deploy farm solutions to SharePoint farms and Version them. I'd like to integrate my work in xSharepPoint. The

msftclas commented:
Hi __@wulfland__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@wulfland__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

wulfland commented:
Can someone tell me how I can run the tests locally?

wulfland commented:
I figured it out. I will fix the tests the next days.

BrianFarnhill commented on Modules/xSharePoint/DSCResources/MSFT_xSPFarmSolution/MSFT_xSPFarmSolution.schema.mof:
```diff
@@ -0,0 +1,26 @@
+[ClassVersion("1.0.0.0"), FriendlyName("xSPFarmSolution")]
+class MSFT_xSPFarmSolution : OMI_BaseResource
+{
+    [Key] string Name;
```
@TravisEz13 Yea solution name is the key here, its how they are retrieved with PowerShell and shown in the UI

wulfland replied:
Yes. Only one per farm.

BrianFarnhill commented on Modules/xSharePoint/DSCResources/MSFT_xSPFarmSolution/MSFT_xSPFarmSolution.schema.mof:
```diff
@@ -0,0 +1,26 @@
+    [Write] string Ensure;
+	[Write] string Version;
+	[Write] Boolean Deployed;
+    [Write, EmbeddedInstance("MSFT_Credential")] String InstallAccount;
```
@TravisEz13 Yes, this is by design [as per our guidance](https://github.com/powershell/xSharePoint/wiki/Remote%20sessions%20and%20the%20InstallAccount%20variable) so we can support PowerShell 4 while offering a better approach for PowerShell 5. We have a unit test that makes sure it is optional where it is used.

wulfland commented on Tests/xSharePoint/xSharePoint.xSPFarmSolution.Tests.ps1:
```diff
@@ -0,0 +1,138 @@
+			}
+		}
+	}   
+}
```
@TravisEz13 Is there an easy way to check and fix style Guideline issues? Like Tabs and new line...

BrianFarnhill replied:
@wulfland The tabs/spaces are checked for in a unit test, and so are a few others. We'll be adding tests for others though

BrianFarnhill commented on Modules/xSharePoint/DSCResources/MSFT_xSPFarmSolution/MSFT_xSPFarmSolution.psm1:
```diff
@@ -0,0 +1,341 @@
+
+    start-sleep -s 5
+
+    $result = Invoke-xSharePointCommand -Credential $InstallAccount -Arguments @{ Name = $SolutionName } -ScriptBlock {
```
Have you tested that this works correctly? You aren't passing $InstallAccount through to this function, so you might need to add that (I suspect it's running correctly with that being null and PowerShell 5 being present)

wulfland replied:
I added the account as a parameter.

BrianFarnhill commented on Modules/xSharePoint/DSCResources/MSFT_xSPFarmSolution/MSFT_xSPFarmSolution.schema.mof:
```diff
@@ -0,0 +1,26 @@
+/*
+**Description**
+
+This resource will deploy a farm solution to a farm. A version will be stored in the property bag to determin later if the correct version is installed.
```
We need these descriptions to be markdown compliant as well because we generate wiki pages with them, so can you update this to start each sentence on a new line?

BrianFarnhill commented on Tests/xSharePoint/xSharePoint.xSPFarmSolution.Tests.ps1:
```diff
@@ -0,0 +1,138 @@
+				WebApplications = @()
+			}
+
+			Mock Get--TargetResource { return $actualValues }
```
There are two dashes in the mock name here instead of one

BrianFarnhill commented on Tests/xSharePoint/xSharePoint.xSPFarmSolution.Tests.ps1:
```diff
@@ -0,0 +1,138 @@
+        
+        Import-Module $Global:CurrentSharePointStubModule -WarningAction SilentlyContinue
+
+		Context "When the solution does not exist,"{
```
When we are creating tests and the contexts for them, we try to describe it as "the scenario is this, and should be this" - so for example:

 - The solution isn't installed, but should be
 - The solution isn't installed, and should not be

Then from there we would have "it" statements to test each of the three resources, so things like "it returns 0 from the get method", "it returns false from the test method", "the set method runs the install function" and things like that. It gives us a good structure to test for the potential states a system could be in, and validate the logic we have in each function. So are you able to have a look at the structure you have for the tests here? This first context for example you are essentially testing each property in its own test, and not really describing what the desired state of the context is either. Happy to help with it if that doesn't make sense though - let me know though.

wulfland commented:
I added a first test Scenario in your style. Please leave a small comment, that this is ok. Then I will add the other scenarions. I also Need to test the module in a farm enviromnet, The logic works and is well tested - but as a script with anual get/test/set calls - and not as a module. I'll let you know.

ykuijs commented on Modules/xSharePoint/DSCResources/MSFT_xSPFarmSolution/MSFT_xSPFarmSolution.schema.mof:
```diff
@@ -0,0 +1,27 @@
+class MSFT_xSPFarmSolution : OMI_BaseResource
+{
+    [Key] string Name;
+    [Required] string LiteralPath;
```
Haven't tested the resource, but do have a question: Does the literal path also accept UNC paths??

wulfland replied:
Yes. The Add-SPSolution CmdLet also accepts UNC AFAIK.

ykuijs commented on Modules/xSharePoint/DSCResources/MSFT_xSPFarmSolution/MSFT_xSPFarmSolution.schema.mof:
```diff
@@ -0,0 +1,27 @@
+{
+    [Key] string Name;
+    [Required] string LiteralPath;
+    [Write] string WebApplications[];
```
What about solutions that need to be Globally Deployed? Is this also possible?

wulfland replied:
Yes. See line 218 to 236. You can specify one or more web application. If you do not specify any it defaults to AllWebApplications if it has "ContainsWebApplicationResource". Global Solutions are installed without the WebApplication Parameter.

wulfland commented:
Thanks @BrianFarnhill How much code coverage is required? I have now 76%.

wulfland commented on README.md:
```diff
@@ -98,6 +98,8 @@ Additional detailed documentation is included on the wiki on GitHub.
 
  * Added xSPUserProfileSyncConnection
 
+ * Added xSPFarmSolution resource
```
You mean in the Header of the mof file, right? Haven't found anoterh place where the resources are documented.

wulfland commented:
There is no readm.mof. Only a readme.md. In there I added the resource to unreleased. Anything else I have to do?

wulfland commented:
I added the resource to the unreleased section. I can't find more description. Think that should be it.

### Unreleased 
 * Added xSPWordAutomationServiceApp resources
 * Added xSPHealthAnalyzerRuleState resources
 * Added xSPUserProfileProperty
 * Added xSPWorkManagementApp 
 * Added xSPUserProfileSyncConnection
 * **Added xSPFarmSolution resource**

BrianFarnhill commented:
@TravisEz13 I've normally been adding them when I do a release but wouldn't say no to it going in here. I actually think I wanna automate that list now that it's getting so big but thats another issue

@wulfland He is just talking about the list of all resources in the module, that is above the unreleased section that you already added to, just so your resource is there. Also though, can I get you to add a heading "unreleased" above the v0.10 heading and put your item in there? We released v0.10 already so this will go out in v0.11 so it needs to not be under the v0.10 heading. Thanks dude

wulfland commented:
Thats how I thought the workflow is. I add the resource to the list under unreleased and you add it to the release notes and the global list if you create a release.
I hope I've got it right this time...

BrianFarnhill commented on Modules/xSharePoint/DSCResources/MSFT_xSPFarmSolution/MSFT_xSPFarmSolution.psm1:
```diff
@@ -0,0 +1,341 @@
+function Set-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Boolean])]
```
Can you remove the output type here - we expect no output from the set method normally

BrianFarnhill commented:
One minor comment I just made and a fresh pull from dev so its mergable then I can merge this one in, I'm happy with everything else

wulfland commented:
Ready to go...

BrianFarnhill commented:
Awesome - and the tests still looks good. All good from my end, merging now. Thanks for the contribution @wulfland !
